chainstate/tx-verifier/src/transaction_verifier/utxos_undo_cache/mod.rs

```diff
@@ -0,0 +1,258 @@
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use super::{error::ConnectTransactionError, CachedOperation, TransactionSource};
+use common::{chain::Transaction, primitives::Id};
+use utxo::{UtxosBlockRewardUndo, UtxosTxUndo, UtxosTxUndoWithSources};
+
+mod cached_utxo_block_undo;
+pub use cached_utxo_block_undo::CachedUtxosBlockUndo;
+
+pub type CachedUtxoBlockUndoOp = CachedOperation<CachedUtxosBlockUndo>;
+
+/// Struct that can hold utxo undo data for blocks/mempool.
+///
+/// There is now restriction on how many blocks a TransactionVerifier can be used on,
+/// so this struct has to work with an arbitrary number of items.
+/// Furthermore every block consists of an arbitrary number of transactions. So one of the jobs here
+/// is to initialize BlockUndo structure when the first transaction arrives and erase
+///
+/// On every level of its nested structure the `CachedOperation` is used to keep the information
+/// about whether data should be eventually added or erased from the db after hierarchy of verifiers
+/// fold.
+#[derive(Debug, Eq, PartialEq)]
+pub struct UtxosBlockUndoCache {
+    data: BTreeMap<TransactionSource, CachedUtxoBlockUndoOp>,
+}
+
+impl UtxosBlockUndoCache {
+    pub fn new() -> Self {
+        Self {
+            data: BTreeMap::new(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(data: BTreeMap<TransactionSource, CachedUtxoBlockUndoOp>) -> Self {
+        Self { data }
+    }
+
+    pub fn data(&self) -> &BTreeMap<TransactionSource, CachedUtxoBlockUndoOp> {
+        &self.data
+    }
+
+    pub fn consume(self) -> BTreeMap<TransactionSource, CachedUtxoBlockUndoOp> {
+        self.data
+    }
+
+    /// Check whether transaction can be disconnected.
+    ///
+    /// An undo object must be available and no other tx must be dependant on the current one.
+    pub fn can_disconnect_transaction<F, E>(
+        &self,
+        tx_source: &TransactionSource,
+        tx_id: &Id<Transaction>,
+        fetcher_func: F,
+    ) -> Result<bool, ConnectTransactionError>
+    where
+        F: Fn(TransactionSource) -> Result<Option<CachedUtxosBlockUndo>, E>,
+        ConnectTransactionError: From<E>,
+    {
+        let block_undo = match self.data.get(tx_source) {
+            Some(op) => match op {
+                CachedOperation::Write(undo) | CachedOperation::Read(undo) => {
+                    Ok(Some(undo.clone()))
+                }
+                CachedOperation::Erase => Ok(None),
+            },
+            None => fetcher_func(*tx_source).map_err(ConnectTransactionError::from),
+        }?;
```

Comment:
Tiny nitpick: I'd probably place the question mark operator after `map_err(...)`. It's closer to the source of the error, so maybe a little easier to see what's going on and you don't need the extra `Ok` wrappers in the `Some` branch.